pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -253,8 +295,210 @@ void Stencil::CollectStrideOneIndices() {
+  SearchIndex(0);
+  // B A C, Search M(0) first as M is most restricted index
+  tensorsOrder[0] = 1;
+  tensorsOrder[1] = 0;
+  SearchIndex(0);
+}
+
+void Stencil::SearchTiles(unsigned idx) {
+  if (idx >= kNumIndex) {
+    double performance = Evaluate();
+    if (performance < bestPerf) {
+      bestPerf = performance;
+      for (unsigned i = 0; i < kNumTensors; ++i) {
+        bestTensorsOrder[i] = tensorsOrder[i];
+      }
+      for (unsigned i = 0; i < kNumIndex; ++i) {
+        bestIdxs[i] = innerIdxs[i];
+        bestTiles[i] = tiles[i];
+      }
+    }
+    return;
+  }
+
+  unsigned range = idxRange(innerIdxs[idx]);
+  for (unsigned i = range; i > 0; --i) {
+    if (onlyEven[idx] && (range % i != 0)) {
+      continue;
+    }
+    tiles[idx] = i;
+    SearchTiles(idx + 1);
+  }
+}
+
+int64_t Stencil::idxRange(mlir::BlockArgument idx) {
+  auto pf = mlir::cast<mlir::AffineParallelOp>(idx.getOwner()->getParentOp());
+  auto ranges = pf.getConstantRanges();
+  if (ranges != llvm::None) {
+    return (*ranges)[idx.getArgNumber()];
+  }
+  return -1;
+}
+
+double Stencil::Evaluate() {
+  unsigned tot_inner_loop = tiles[0] * tiles[1] * tiles[2];
+  double throughput;
+  unsigned startup_cost;
+  std::tie(throughput, startup_cost) = Throughput(tiles[0], tiles[1], tiles[2]);
+  if (throughput == 0) {
+    return std::numeric_limits<double>::max();
+  }
+  double inner_time = tot_inner_loop / throughput;
+  IVLOG(3,
+        "Inner: loop = " << tot_inner_loop << " inner_time = " << inner_time);
+  for (unsigned i = 0; i < kNumIndex; ++i) {
+    IVLOG(3, innerIdxs[i] << ": " << tiles[i]);
+  }
+
+  llvm::DenseMap<mlir::BlockArgument, unsigned> middle_idxs;
+  for (auto idx : accIdxs) {
+    middle_idxs.try_emplace(idx, idxRange(idx));
+  }
+  for (unsigned i = 0; i < kNumIndex; ++i) {
+    auto it = middle_idxs.find(innerIdxs[i]);
+    if (it != middle_idxs.end()) {
+      it->second = (it->second - 1) / tiles[i] + 1;
+    }
+  }
+  unsigned tot_middle_loop = 1;
+  for (auto &kvp : middle_idxs) {
+    tot_middle_loop *= kvp.second;
+  }
+
+  IVLOG(3, "Middle: loop = " << tot_middle_loop);
+
+  for (auto &kvp : middle_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, kvp.first << ": " << kvp.second);
+    }
+  }
+
+  llvm::DenseMap<mlir::BlockArgument, unsigned> outer_idxs;
+  for (auto idx : outIdxs) {
+    outer_idxs.try_emplace(idx, idxRange(idx));
+  }
+  for (unsigned i = 0; i < kNumIndex; ++i) {
+    auto it = outer_idxs.find(innerIdxs[i]);
+    if (it != outer_idxs.end()) {
+      it->second = (it->second - 1) / tiles[i] + 1;
+    }
+  }
+  unsigned tot_outer_loop = 1;
+  for (auto &kvp : outer_idxs) {
+    tot_outer_loop *= kvp.second;
+  }
+
+  IVLOG(3, "Outer: loop = " << tot_outer_loop);
+
+  for (auto &kvp : outer_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, kvp.first << ": " << kvp.second);
+    }
+  }
+
+  unsigned outer_batches =
+      (tot_outer_loop - 1) / std::thread::hardware_concurrency() + 1;
```

Comment:
The number of threads should be a parameter, not something that we probe on the current machine. That's because we want to be able to support AOT use cases.

Reply:
Added a way to set it internally. Need to decide on passing the parameter from the outside. I'd say using a compiler option, but we can discuss this.

Reply:
We should try to use instance-specific options over global ones, see this: https://mlir.llvm.org/docs/WritingAPass/#instance-specific-pass-options

For testing it's probably not a big deal to use global options, provided there's a way to configure the option programmatically from a pipeline definition.

Reply:
Added code to use thye instance options.

Reply:
This still seems to use the `std::thread::hardware_concurrency()`, is this intentional?